File: permutations/permutations.py
def helper(index,nums,res):
    if index==len(nums):
        res.append(nums[:])
        return
    for i in range(index,len(nums)):
        nums[i],nums[index]=nums[index],nums[i]
        helper(index+1,nums,res)
        nums[i],nums[index]=nums[index],nums[i]

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res=[]
        helper(0,nums,res)
        return res
        

File: permutations/permutations.cpp
class Solution {
    vector<vector<int>>ans;
    void f(int cur, vector<int>nums){
        if (cur==nums.size()){
            ans.push_back(nums);
            return;
        }
        for (int i=cur; i<nums.size(); i++){
            swap(nums[cur],nums[i]);
            f(cur+1,nums);
            swap(nums[cur],nums[i]);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        f(0,nums);
        return ans;
    }
};